feat: cache get-app artifacts by commit_hash

18alantom opened the pull request:
TLDR: Stores `get-app` artifacts in `.cache` by ~commit hash~ cache key and reuses them when applicable.

Partially Fixes: https://github.com/frappe/press/issues/1242

**TODO:**
- [x] Cache output and restore cache if present.
- [x] Add explicit `compress_artifacts` flag.
- [x] Update `frappe/esbuild/esbuild.js` to account for cached get-app. https://github.com/frappe/frappe/pull/24412
- [x] Complete installation from cache.
- [x] Ensure `$BENCH/sites` is being updated properly.
- [x] Get rid of `node_modules` for apps with built frontends (gameplan, insights, etc).
- [x] Commands to clear, and view app cache.
- [x] Check if it works.

### Usage

Adds two new options to `bench get-app`:
- `--cache-key`: Used to store tarred app folder in `~/.cache/bench/apps` with the following format `"{app}-{cache-key[:10]}.tar"`
- `--compress-artifacts`: Optional flag that causes artifacts to be compressed using gzip, extension will then be `.tgz`.

**Examples:**

```bash
# Get from local repo, apps/gampelan will be cached if not present
bench get-app file:///home/frappe/gameplan --cache-key 87d4aa3b10aa23a620d5a51a968ae47110c43dd1

# Get from remote frappe repo, apps/erpnext will be compressed and cached if not present
bench get-app erpnext --cache-key 7bcea6099da1a41ad530bdd85ff337f627039b80 --compress-artifacts
```

For either of the above if app is present in cache it will be fetched from there, else it will be stored in cache. While installing, cache is checked for both compressed and non compressed tar files.

### Utility Commands

<details>
<summary><code>bench app-cache</code></summary>


<img width="792" alt="Screenshot 2024-01-19 at 19 10 18" src="https://github.com/frappe/bench/assets/29507195/ea021d33-35da-4563-94a6-c5e3e9133053">


</details>


<details>
<summary><code>bench app-cache --remove-app APP_NAME</code></summary>

<img width="792" alt="Screenshot 2024-01-19 at 19 10 40" src="https://github.com/frappe/bench/assets/29507195/b6e48ce8-ffb3-484e-9aa5-6b70506b0c8b">

</details>

<details>
<summary><code>bench app-cache --remove-key CACHE_KEY</code></summary>

<img width="792" alt="Screenshot 2024-01-19 at 19 11 23" src="https://github.com/frappe/bench/assets/29507195/56abb37f-127c-478e-9a61-98d5f977aed8">

</details>

### Why?

When running `get-app` bench spends significant amount of time in git clone, yarn install, pip install, building frontend dependencies. If it's running get-app on the same code (same app, same repo, same commit hash), `get-app` artifacts from the previous run can be reused.


### How?

After `get-app` has completed for some `$APP` the relevant folders that change are:
- `$BENCH/apps/$APP`: cloned source files, and built frontend files and dependencies.
- `$BENCH/env`: python dependencies for the get-app'd `$APP`.
- `$BENCH/sites`: link to the `assets/$APP` folder, update `assets.json`, `apps.json`, `apps.txt`.

Caching the contents of `$BENCH/apps/$APP` covers repo cloning, yarn install and yarn build. That's what this PR facilitates.

<details>
<summary>Reference</summary>

File system changes after running `get-app` for `frappe/hrms`.

<img width="751" alt="Screenshot 2024-01-15 at 11 49 18" src="https://github.com/frappe/bench/assets/29507195/74c7260e-a394-41a5-b6dc-2888972a3ee0">

</details>

**Regarding `$BENCH/env`**

Pip install does take a considerable amount of time but since it isn't contained in the app folder it won't be included for now. Future updates might have python dependencies stored inside the app folder. Out of scope for now.

18alantom committed:
feat: cache get-app artifacts by commit_hash

18alantom committed:
fix: prevent circular dependency
- put check_existing_dir back in its place

18alantom committed:
fix: update install_app with using_cached flag

ankush commented:
>  Get rid of node_modules for apps with built frontends (gameplan, insights, etc).

This will not allow rebuild btw (we allow doing it from FC dashboard :grimacing: )

ankush commented on bench/app.py:
```diff
@@ -283,6 +290,90 @@ def update_app_state(self):
+			message += " (compressed)"
+		click.secho(message)
+
+		os.chdir(app_path.parent)
```
If possible avoid changing directories. This has caused problems in past as bench assumes a certain directory to be cwd always.

18alantom replied:
Here it's required cause of how tar functions, it stores the entire path and untarring requires moving the file out of nested directories. I've wrapped tar in a `try` `finally` so `cwd` should be restored irrespective: https://github.com/frappe/bench/pull/1519/commits/30f301e3ffa8a2561160ec740d2d046355905c97

18alantom commented:
> > Get rid of node_modules for apps with built frontends (gameplan, insights, etc).
> 
> This will not allow rebuild btw (we allow doing it from FC dashboard 😬 )

Ah, I didn't know we allowed rebuilds. I suppose we can trigger `yarn install` if `node_modules` not found. It will increase the first rebuild time, but seems like a reasonable tradeoff. What do you think?

18alantom committed:
fix: check if installed FF supports use-cached

18alantom committed:
refactor: use env to trigger using-cached flag

Click supports pulling args from an envvar if it is present,
this would be quicker and cleaner than calling a dummy command
to check if the feature is supported

18alantom committed:
fix: remove unused nodemodules before caching

18alantom committed:
fix: wrap tarfile with error handling
- ensure return to cwd after tarring

18alantom committed:
feat: add bench app-cache helper command

18alantom committed:
chore: rename commit-hash to cache-key

18alantom committed:
fix: remove rem check from app-cache --clear

18alantom committed:
chore: bump Jinja appease CI

18alantom commented:
**Note: ignoring SonarCloud Quality Gate**

It's failing due to potential tar bombs. It's failed in two places when creating the archive and when extracting it.

**When creating an archive**: check is not necessary here cause the archive file is newly created and will be empty. Else that stack of code would not have run.

**When extracting from an archive**: archive being extracted from is created by bench, and not user submitted, well the app repo could be, but in the given context it seems highly unlikely.

(that being said I might add an `extractall` filter in a subsequent PR, it is not going to appease SonarCloud but will address the issue)


_Edit: filter added in https://github.com/frappe/bench/pull/1523_

18alantom committed:
Merge branch 'develop' into get-app-cache
